src/core/lib/event_engine/grpc_ares_wrapper.cc

```diff
@@ -0,0 +1,1013 @@
+    EventEngine* event_engine)
+    : GrpcAresRequest(name, absl::nullopt, timeout,
+                      std::move(polled_fd_factory), event_engine) {}
+
+GrpcAresSRVRequest::~GrpcAresSRVRequest() {
+  GRPC_ARES_WRAPPER_TRACE_LOG("request:%p destructor", this);
+}
+
+void GrpcAresSRVRequest::Start(
+    absl::AnyInvocable<void(absl::StatusOr<Result>)> on_resolve) {
+  auto self = Ref(DEBUG_LOCATION, "Start");
+  grpc_core::MutexLock lock(&mu_);
+  GPR_ASSERT(initialized_);
+  // Don't query for SRV records if the target is "localhost"
+  if (gpr_stricmp(std::string(host_).c_str(), "localhost") == 0) {
+    event_engine_->Run([on_resolve = std::move(on_resolve)]() mutable {
+      grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
+      grpc_core::ExecCtx exec_ctx;
+      on_resolve(GRPC_ERROR_CREATE(
+          "Skip querying for SRV records for localhost target"));
+    });
+    Unref(DEBUG_LOCATION, "Start");
+    return;
+  }
+  on_resolve_ = std::move(on_resolve);
+  // Query the SRV record
+  service_name_ = absl::StrCat("_grpclb._tcp.", host_);
+  ares_query(channel_, service_name_.c_str(), ns_c_in, ns_t_srv,
+             &GrpcAresSRVRequest::OnSRVQueryDoneLocked,
+             static_cast<void*>(this));
+  if (!shutting_down_) {
+    Work();
+    StartTimers();
+  }
+}
+
+void GrpcAresSRVRequest::OnResolve(absl::StatusOr<Result> result) {
+  auto closer =
+      absl::MakeCleanup([this] { Unref(DEBUG_LOCATION, "OnResolve"); });
+  if (cancelled_) {
+    // Cancel does not invoke on_resolve.
+    return;
+  }
+  shutting_down_ = true;
+  CancelTimers();
+  event_engine_->Run([on_resolve = std::move(on_resolve_),
+                      result = std::move(result)]() mutable {
+    grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
+    grpc_core::ExecCtx exec_ctx;
+    on_resolve(std::move(result));
+  });
+}
+
+GrpcAresTXTRequest::GrpcAresTXTRequest(
+    absl::string_view name, EventEngine::Duration timeout,
+    std::unique_ptr<GrpcPolledFdFactory> polled_fd_factory,
+    EventEngine* event_engine)
+    : GrpcAresRequest(name, absl::nullopt, timeout,
+                      std::move(polled_fd_factory), event_engine) {}
+
+GrpcAresTXTRequest::~GrpcAresTXTRequest() {
+  GRPC_ARES_WRAPPER_TRACE_LOG("request:%p destructor", this);
+}
+
+void GrpcAresTXTRequest::Start(
+    absl::AnyInvocable<void(absl::StatusOr<Result>)> on_resolve) {
+  auto self = Ref(DEBUG_LOCATION, "Start");
+  absl::ReleasableMutexLock lock(&mu_);
+  GPR_ASSERT(initialized_);
+  // Don't query for TXT records if the target is "localhost"
+  if (gpr_stricmp(std::string(host_).c_str(), "localhost") == 0) {
+    event_engine_->Run([on_resolve = std::move(on_resolve)]() mutable {
+      grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
+      grpc_core::ExecCtx exec_ctx;
+      on_resolve(
+          GRPC_ERROR_CREATE("Skip querying for TXT records localhost target"));
+    });
+    lock.Release();
+    Unref(DEBUG_LOCATION, "Start");
```

Comment:
I believe this will double-unref the request. You probably want the ref to be `Ref(...).release()`, to maintain the ref through the `Done` callback.